### ISSUE TITLE
[L] API users should be able to merge side tags

### ISSUE THREAD

**Issue by bowlofeggs** [fedora-infra/bodhi#2319]
We need to add the ability for Bodhi to merge side tags. Bodhi should make sure the tests pass before merging, and it should also ensure there will be no conflicts due to the merge. Once the side tag update is merged, the update should either progress to stable (for Rawhide), or should go to updates-testing (for other releases).

If the merge fails, Bodhi should ensure that the packager has the information they need to resolve the conflict and try again. We will need to watch their side tag for new or removed builds and update the update accordingly by using a fedora-messaging consumer.

**Comment by nphilipp** [comment]
@bowlofeggs:

> Bodhi should make sure the tests pass before merging, and it should also ensure there will be no conflicts due to the merge.

Is this functionality in Bodhi already, or are there issues/cards for tests passing and no merge conflicts?

> We will need to watch their side tag for new or removed builds and update the update accordingly by using a fedora-messaging consumer.

This sounds like it could be its own task, what do you think?

**Comment by bowlofeggs** [comment]
On Mon, 2019-05-27 at 06:14 -0700, Nils Philippsen wrote:
> > Bodhi should make sure the tests pass before merging, and it should
> > also ensure there will be no conflicts due to the merge.
> 
> Is this functionality in Bodhi already, or are there issues/cards for
> tests passing and no merge conflicts?

Hey Nils!

We do have a "do the tests pass" in Bodhi already (the
Update.test_gating_status field), but we do not have any code for
merging things in Koji or detecting conflicts.

> > We will need to watch their side tag for new or removed builds and
> > update the update accordingly by using a fedora-messaging consumer.
> 
> This sounds like it could be its own task, what do you think?

Sure, we could split it up if you prefer. Feel free to file another
issue for this part of it and put it on the backlog. If you do so,
please link it here as well.

**Comment by cverna** [comment]
As a general note merging tag sounds like something we want to do in a celery tasks, so this ticket really depends on #2851

**Comment by cverna** [comment]
The script used by releng to merge side tags is (https://pagure.io/releng/blob/master/f/scripts/mass-tag.py)

**Comment by mohanboddu** [comment]
I want to bring up a use case here:

If there is package `foo` and maintainer A requests a side tag and builds the `foo-1.1` in the side tag. Then maintainer B builds the same package `foo-1.2` and sends it as a single update or multiple update before maintainer A. But when maintainer A tries to merge the side tag, there might be a conflict of which build to trust or both might be bad because both might be solving different issues.

What I would like to suggest is:
When maintainer A requests a side tag, bodhi should record the time when the tag is created. And when maintainer A pushes the update, bodhi should know that there is a `foo-1.2` build which is in rawhide after the side tag is created, then bodhi should give the maintainer A an option to pick which one is the best one. If he picks `foo-1.1` then it will get tagged and becomes the latest, if he picks `foo-1.2` bodhi should not do anything about `foo-1.1` build. Or he might rebuild it again `foo-1.3` and bodhi will now give two options `foo-1.2` or `foo-1.3` and he can pick `foo-1.3` which has fixes from both `1.1` and `1.2`

**Comment by nphilipp** [comment]
For reference, here's pypingou's multi-build update workflow diagram \[[download](https://user-images.githubusercontent.com/820624/63262799-34940480-c287-11e9-9594-e177ce7e6d4d.png)\]:

![Multi-builds - Rawhide Package Gating Flow.png](https://user-images.githubusercontent.com/820624/63262799-34940480-c287-11e9-9594-e177ce7e6d4d.png)

I'm taking this card to be the overarching one for the whole workflow, not just the last step in the diagram. :smiley:

**Comment by nphilipp** [comment]
In Robosignatory's [PR#27](https://pagure.io/robosignatory/pull-request/27) I added tests around its "signing a tagged build" workflow, which I wanted to get in in order not to screw things up (too much).

**Comment by nphilipp** [comment]
Here are the "sub-cards" covering individual bits and pieces in Bodhi and robosig:

* #3473 Update from side tag: create appropriate auxiliary tags and tag builds accordingly
* [robosig/#28](https://pagure.io/robosignatory/issue/28) CI Gating: Robosignatory should sign Koji builds from side-tags
* #3475 Message consumer: Move update from side-tag to testing
* #3476 Message consumer: Move tested update from testing to stable and remove side and aux tags.

**Comment by nphilipp** [comment]
Current state:

* @Zlopez is working on #3473, #3475
* I've implemented [robosig/#28](https://pagure.io/robosignatory/issue/28) in [robosig/#29](https://pagure.io/robosignatory/pull-request/29) which @abompard reviewed (merge pending comment from Patrick)
* I've implemented #3476 in PR #3489 (but it needs more eyeballs)

**Comment by cverna** [comment]
I think we can close this since all the related tasks have been done.
